src/main/java/com/fasterxml/jackson/core/json/UTF8JsonGenerator.java

```diff
@@ -1510,7 +1515,14 @@ private final void _writeStringSegment2(final char[] cbuf, int offset, final int
                 outputBuffer[outputPtr++] = (byte) (0xc0 | (ch >> 6));
                 outputBuffer[outputPtr++] = (byte) (0x80 | (ch & 0x3f));
             } else {
-                outputPtr = _outputMultiByteChar(ch, outputPtr);
+                // multibyte character
+                if (Feature.WRITE_UTF8_SURROGATES.enabledIn(_features) && isSurrogatePair((char) ch) && offset < end) {
+                    char highSurrogate = (char) ch;
```

Comment:
it might be possible to create a class variable to hold the result of Feature.WRITE_UTF8_SURROGATES.enabledIn(_features) so that we don't need to recalculate this for every char.